File: go/grep.go
package grep

import (
	"fmt"
	"os"
	"regexp"
	"strings"
)

func Search(pattern string, flags, files []string) []string {
	showLineNum := false
	onlyFilename := false
	caseInsensitive := false
	invert := false
	matchEntireLine := false

	for _, flag := range flags {
		switch flag {
		case "-n":
			showLineNum = true
		case "-l":
			onlyFilename = true
		case "-i":
			caseInsensitive = true
		case "-v":
			invert = true
		case "-x":
			matchEntireLine = true
		}
	}

	if caseInsensitive {
		pattern = "(?i)" + pattern
	}

	if matchEntireLine {
		pattern = "^" + pattern + "$"
	}

	re, err := regexp.Compile(pattern)
	if err != nil {
		return []string{}
	}

	res := []string{}
	for _, file := range files {
		content, _ := os.ReadFile(file)
		lines := strings.Split(string(content), "\n")
		for i, line := range lines {
			line = strings.Trim(line, " ")
			if line == "" {
				continue
			}

			match := re.MatchString(line)
			if match && !invert || !match && invert {
				if onlyFilename {
					res = append(res, file)
					break
				}
				if showLineNum {
					line = fmt.Sprintf("%d:%s", i+1, line)
				}

				if len(files) > 1 {
					line = file + ":" + line
				}

				res = append(res, line)
			}
		}
	}
	return res
}

File: elixir/grep.ex
defmodule Grep do  
  @spec grep(String.t(), [String.t()], [String.t()]) :: String.t()
  def grep(pattern, flags, files) do
    pattern = if "-x" in flags, do: "^#{pattern}$", else: pattern
    opts = if "-i" in flags, do: "i", else: ""
    r = Regex.compile!(pattern, opts)
    show_filename = length(files) > 1
    show_linenum = "-n" in flags
    invert = "-v" in flags
    
    files
    |> Enum.map(fn file ->
      content = File.read!(file)
      if "-l" in flags do
        if xor(Regex.match?(r, content), invert) do
          "#{file}\n"
        else
          ""
        end
      else
        content
        |> String.trim("\n")
        |> String.split("\n")
        |> Enum.with_index(1)
        |> Enum.reduce("", fn {line, i}, acc ->
          if xor(Regex.match?(r, line), invert) do
            line = if show_linenum, do: "#{i}:#{line}", else: line
            line = if show_filename, do: "#{file}:#{line}", else: line
            acc <> "#{line}\n"
          else
            acc
          end
        end)
      end
    end)
    |> Enum.join("")
  end

  defp xor(true, false), do: true
  defp xor(false, true), do: true
  defp xor(_, _), do: false
end

